### PR TITLE
push to dockerhub

### PR DESCRIPTION
**Pull request by jburel**
This PR aims at removing a manual step of the openmicroscopy release process
When the repository is tagged, a docker image is built and pushed

Secrets will need to be added.
I will roll out to other repo when we are happy with the approach.
cc @joshmoore @sbesson 

See https://hub.docker.com/repository/docker/jburel/omero-server-docker

### CODE REVIEW

**Comment by sbesson** [comment]
Looks good as a starting step, the building and pushing is something I have been hoping we could automate through GH actions. A couple of immediate design thoughts:

- Docker Hub vs GitHub Container Registry: would there be value in pushing to both places for redundancy?
- I see that `docker/build-push-action` has now a `v2` of the action which seems to deprecate the `tag_with_ref` input (with a more complex logic) - have you looked at whether it would be worth upgrading?
- the previous release workflow was `make VERSION=x.y.z git-push &&  make VERSION=x.y.z docker-build  &&  make VERSION=x.y.z docker-push`. This included the build and push of multiple tags of the same Docker image. How is that modified by this workflow?

**Comment by jburel** [comment]
* I am not convinced that there is much value to push to 2 different places similar to artifactory vs sonatype
* I have not looked at v2 yet.
* ``git-push`` and ``docker-build/push`` are 2 different things. ``git-push`` will be pushing a tag to the repo, we will still need that step to trigger any action. The ``docker`` part is build/push via the action. What is missing with in this PR, is the fact we don't have a latest, x.y so

**Comment by manics** [comment]
Since I'm still getting notifications here... I wrote https://github.com/manics/action-major-minor-tag-calculator a while ago to handle the Docker tags use case. Use at your own risk 😄

**Comment by jburel** [comment]
I have pushed a new commit using v2 and @manics action
I also put the logic in place so we can change the name of the prefix (this repo uses ``omero-server`` for example) or use the name of the repository in the prefix https://hub.docker.com/repository/docker/jburel/omero-server

**Comment by manics** [comment]
Looks like `github.context.payload.ref` isn't defined for PRs
https://github.com/manics/action-major-minor-tag-calculator/blob/98300ea16dac344e9973e1ca9a6c509cb8454c86/src/index.js#L17
(the branch passes https://github.com/jburel/omero-server-docker/runs/1832767447)

**Comment by jburel** [comment]
@manics I have split the workflow into 2 jobs since the upload will happen on tag
so hopefully that should work

**Comment by jburel** [comment]
@manics I had to make some adjustments to do what we had in-place when pushing images to docker.
It now does what we want e.g. tagging 5.6.x-y now will return 5.6.x-y, 5.6.x, 5.6, 5 and latest (assuming it is the latest in that example).
What about transferring the repo ``action-major-minor-tag-calculator`` to the organisation and we add you as a maintainer?

**Review by sbesson — APPROVED** [review]
As a general rule, moving towards automated DockerHub deployment on tag has been something we have been wanting to achieve for a while. It looks like the current workflow is matching all requirements, consuming a `x.y.z[-build]` Git tag and turning it into a series of Docker tags that can be pushed to a registry.

Proposing to get this merged so that we can test the logic (assuming we will use the `omereleases` agent credentials as secrets) and propagate to other repositories as required

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -14,3 +14,36 @@ jobs:
...
+          prefix: "${{ steps.getprefix.outputs.prefix }}"
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Push to Docker Hub
+        uses: docker/build-push-action@v2
```
Comment (sbesson) [inline .github/workflows/main.yml:46]: As a follow-up, we could take advantage of the action `labels` property (see https://github.com/docker/build-push-action/blob/master/docs/advanced/tags-labels.md) and start generating the metadata that's currently hardcoded and updated via https://github.com/ome/omero-server-docker/blob/eba93585868447eb84f84d3582106cf4dae7d455/Makefile#L36-L38 as part of the build process

**Pull request merged**